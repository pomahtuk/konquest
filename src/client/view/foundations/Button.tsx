import React, { ReactElement, useContext, ReactNode } from "react";
import styled from "@emotion/styled";
import { ThemeContext } from "../themes/ThemeProvider";

type ButtonProps = {
  children?: ReactNode;
  variant?: "primary" | "secondary" | "destructive";
  size?: "large";
  wide?: boolean;
  onClick?: () => void;
  className?: string;
  type?: "button" | "submit";
  disabled?: boolean;
};

const Button = ({ children, type = "button", variant, size, disabled, wide, className, onClick }: ButtonProps): ReactElement => {
  const theme = useContext(ThemeContext);

  const color = variant === "secondary" ? theme.colors.action : theme.colors.white;
  const backgroundColorHover =
    variant === "secondary" ? theme.colors.actionLighter : variant === "destructive" ? theme.colors.destructiveDark : theme.colors.actionDark;
  const backgroundColor = variant === "secondary" ? theme.colors.white : variant === "destructive" ? theme.colors.destructive : theme.colors.action;
  const borderColorHover = variant === "destructive" ? theme.colors.destructiveDark : theme.colors.actionDark;
  const borderColor = variant === "destructive" ? theme.colors.destructive : theme.colors.action;

  const ButtonTag = styled.button`
    border: 1px solid ${borderColor};
    border-radius: ${theme.units.smaller};
    display: inline-flex;
    align-items: center;
    justify-content: center;
    padding: ${theme.units.medium} ${theme.units.large};
    cursor: pointer;
    font-family: inherit;
    text-decoration: none;
    box-sizing: border-box;
    vertical-align: middle;
    position: relative;
    background-color: ${backgroundColor};
    color: ${color};
    font-size: ${theme.fontSizes.small};
    font-weight: ${theme.fontWeights.medium};
    line-height: ${theme.lineHeights.small};

    :hover,
    :active {
      background-color: ${backgroundColorHover};
      border-color: ${borderColorHover};
    }

    [disabled],
    [disabled]:hover {
      background-color: ${theme.colors.grayscale};
      border-color: ${theme.colors.grayscale};
      cursor: not-allowed;
    }

    ${size === "large"
      ? `
      padding: calc(${theme.units.larger} / 2) ${theme.units.larger};
      font-size: ${theme.fontSizes.medium};
      font-weight: ${theme.fontWeights.normal};
      line-height: ${theme.lineHeights.medium};
    `
      : ""}
    ${wide ? "width: 100%;" : ""}
  `;
  return (
    <ButtonTag className={className} onClick={onClick} type={type} disabled={disabled}>
      {children}
    </ButtonTag>
  );
};

export default Button;
