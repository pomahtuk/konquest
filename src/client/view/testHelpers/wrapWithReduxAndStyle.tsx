import React, { ReactElement } from "react";
import { Provider } from "react-redux";
import storeCreator from "../../stores/game.store";
import ThemeProvider from "../themes/ThemeProvider";
import baseTheme from "../themes/base.theme";
const store = storeCreator();

export const CurrentStore = store;

const wrapWithReduxAndStyle = (element: ReactElement): ReactElement => {
  return (
    <Provider store={store}>
      <ThemeProvider theme={baseTheme}>{element}</ThemeProvider>
    </Provider>
  );
};

export default wrapWithReduxAndStyle;
