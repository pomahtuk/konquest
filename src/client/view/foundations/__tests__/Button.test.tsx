import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { matchers } from "jest-emotion";

expect.extend(matchers);

import Button from "../Button";
import wrapWithReduxAndStyle from "../../testHelpers/wrapWithReduxAndStyle";
import baseTheme from "../../themes/base.theme";

const buttonText = "hello";

describe("<Button />", (): void => {
  it("renders <Button /> component", (): void => {
    const { container } = render(wrapWithReduxAndStyle(<Button />));
    expect(container).toMatchSnapshot();
  });

  it("renders <Button /> with passed children", (): void => {
    const { getByText } = render(wrapWithReduxAndStyle(<Button>{buttonText}</Button>));
    expect(getByText(buttonText)).toBeDefined();
  });

  it("renders <Button /> in primary variant", (): void => {
    const { getByText } = render(wrapWithReduxAndStyle(<Button variant="primary">{buttonText}</Button>));
    const element = getByText(buttonText);
    expect(element).toHaveStyleRule("color", baseTheme.colors.white);
    expect(element).toHaveStyleRule("background-color", baseTheme.colors.action);
  });

  it("renders <Button /> in secondary variant", (): void => {
    const { getByText } = render(wrapWithReduxAndStyle(<Button variant="secondary">{buttonText}</Button>));
    const element = getByText(buttonText);
    expect(element).toHaveStyleRule("color", baseTheme.colors.action);
    expect(element).toHaveStyleRule("background-color", baseTheme.colors.white);
  });

  it("renders <Button /> in destructive variant", (): void => {
    const { getByText } = render(wrapWithReduxAndStyle(<Button variant="destructive">{buttonText}</Button>));
    const element = getByText(buttonText);
    expect(element).toHaveStyleRule("color", baseTheme.colors.white);
    expect(element).toHaveStyleRule("background-color", baseTheme.colors.destructive);
  });

  it("renders <Button /> in large size", (): void => {
    const { getByText } = render(wrapWithReduxAndStyle(<Button size="large">{buttonText}</Button>));
    const element = getByText(buttonText);
    expect(element).toHaveStyleRule("font-size", baseTheme.fontSizes.medium);
    expect(element).toHaveStyleRule("padding", `calc(${baseTheme.units.larger} / 2) ${baseTheme.units.larger}`);
  });

  it("renders <Button /> in wide variant", (): void => {
    const { getByText } = render(wrapWithReduxAndStyle(<Button wide>{buttonText}</Button>));
    const element = getByText(buttonText);
    expect(element).toHaveStyleRule("width", "100%");
  });

  it("renders <Button /> in disabled state", (): void => {
    const { getByText } = render(wrapWithReduxAndStyle(<Button disabled>{buttonText}</Button>));
    const element = getByText(buttonText);
    expect(element.getAttribute("disabled")).toBeDefined();
  });

  it("renders <Button /> with additional className", (): void => {
    const { getByText } = render(wrapWithReduxAndStyle(<Button className="some">{buttonText}</Button>));
    const element = getByText(buttonText);
    expect(element.classList.contains("some")).toBe(true);
  });

  it("renders <Button /> with onClick callback", (): void => {
    const fn = jest.fn();
    const { getByText } = render(wrapWithReduxAndStyle(<Button onClick={fn}>{buttonText}</Button>));
    const element = getByText(buttonText);
    fireEvent.click(element);
    expect(fn.mock.calls.length).toBe(1);
  });
});
