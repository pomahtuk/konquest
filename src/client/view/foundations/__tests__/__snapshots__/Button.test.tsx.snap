// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders <Button /> component 1`] = `
.emotion-0 {
  border: 1px solid #0071C2;
  border-radius: 2px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 8px 16px;
  cursor: pointer;
  font-family: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-sizing: border-box;
  vertical-align: middle;
  position: relative;
  background-color: #0071C2;
  color: #fff;
  font-size: 14px;
  font-weight: 500;
  line-height: 20px;
}

.emotion-0:hover,
.emotion-0:active {
  background-color: #005C9D;
  border-color: #005C9D;
}

.emotion-0 [disabled],
.emotion-0 [disabled]:hover {
  background-color: #6B6B6B;
  border-color: #6B6B6B;
  cursor: not-allowed;
}

<div>
  <button
    class="emotion-0"
    type="button"
  />
</div>
`;
